# (1) Project 4 - Web-Scraping

-----
Project 4 requires scraping job data from a a job listings website. I will be scraping the data from au.indeed.com. Once I have scraped and cleaned the data, I will then use the data to try and answer to questions. Firstly, I will make a model to predict the salary of a job. For the second part, using the same scraped data, I will attempt to make a model to predict different job titles to see what features can distinguish between a data science job and a business analyst job.

---

In [1]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

----
I am going to scrape my data from au.indeed.com. I will be scraping data on several features including Job Title, Company, Location, Salary, Job Summary, and Job Description.

I will be scraping data from the 5 main cities in Australia: Sydney, Melbourne, Brisbane, Perth, and Adelaide. I will also be scraping 3 types of job: data scientist, data analyst, and business analyst.

---

## Scraping

In [2]:
indeed_cities = ['Sydney', 'Melbourne', 'Brisbane', 'Perth', 'Adelaide']# Create a list of cities and jobs to scrape.

In [3]:
indeed_jobs = ['data+scientist', 'data+analyst', 'business+analyst']

----
I have created a loop using requests and BeautifulSoup to scrape the data for the different jobs and locations.

---

In [4]:
results = []
desc = []

for job in indeed_jobs:      # Iterate through the jobs list.
    for city in indeed_cities:     # Iterate through the cities list.
        for page in range(0, 1000, 10):   # Iterate through 100 indeed.com pages (page numbers are in increments of 10).
            url = 'https://au.indeed.com/jobs?q=' + job + '&l=' + city + '&start=' + str(page)  # Allows multiple jobs, cities, and pages in the url.
            html = requests.get(url)
            soup = BeautifulSoup(html.text, 'lxml')
            for result in soup.find_all('div', {'class':' row result'}):     # Get data from each job listing.
                results.append(result)    # append results to results list.
            for i in soup.find_all('div', {'class':' row result'}):       # Create another loop to go through the url for each job and scrape the description.
                a = i.find('a')                                                    
                ab = 'https://au.indeed.com/' + a.attrs['href']
                html2 = requests.get(ab)
                soup2 = BeautifulSoup(html2.text, 'lxml')
                desc.append(soup2)      # append results to desc list.

## Functions

---
Now I need to create functions that will allow me to extract the specific data for each feature, e.g. Job Title, Salary, etc.

---

In [5]:
# Job Title.
def get_job(result):   
    try:
        return result.find('a', {'data-tn-element':'jobTitle'}).text    # XPath of job title.
    except:
        return 'NA'      # Return NA if no job title found.

In [6]:
# Company
def get_comp(result):
    try:
        return result.find('span', {'class':'company'}).text
    except:
        return 'NA'

In [7]:
# Location.
def get_loc(result):
    try:
        return result.find('span', {'class':'location'}).text
    except:
        return 'NA'

In [8]:
# Salary.
def get_sal(result):
    try:
        return result.find('span', {'class':'no-wrap'}).text
    except:
        return 'NA'

In [9]:
# Summary.
def get_summ(result):
    try:
        return result.find('span', {'class':'summary'}).text
    except:
        return 'NA'

In [10]:
# Description.
def get_desc(desc):
    try:
        return desc.find('td', {'class' : "snip"}).text
    except:
        return 'NA'

## Dataframe

----
Now I will create two empty dataframes. One for the main data from the job listings (in the results list). The other will have the job description data that was obtained by going through each job listing's url and then scraping (in the desc list). I will then concatonate them.

---

In [11]:
jobs0 = pd.DataFrame(columns=['location', 'title', 'company', 'salary', 'summary'])

In [12]:
jobs1 = pd.DataFrame(columns=['description'])

----
Now I will create 2 for loops to iterate through the scraped data in the results and desc lists and use the functions to add the data to the dataframes.

---

In [13]:
for entry in results:
    location = get_loc(entry)   # use functions to get location data.
    title = get_job(entry)
    company = get_comp(entry)
    salary = get_sal(entry)
    summary = get_summ(entry)
    jobs0.loc[len(jobs0)] = [location, title, company, salary, summary]     # add data to dataframe.

for entry in desc:
    description = get_desc(entry)
    jobs1.loc[len(jobs1)] = [description]    

In [14]:
jobs0 = jobs0.replace(r'\n',' ', regex=True)     # Can use regular expression to remove new lines (\n) from text in the dataframe.

In [15]:
jobs1 = jobs1.replace(r'\n',' ', regex=True)

In [16]:
jobs0.head(15)

location  \
0                 Sydney NSW   
1                 Sydney NSW   
2                 Sydney NSW   
3   Macquarie University NSW   
4                 Sydney NSW   
5                 Sydney NSW   
6                 Sydney NSW   
7             Alexandria NSW   
8                 Sydney NSW   
9                 Sydney NSW   
10                Sydney NSW   
11                Sydney NSW   
12                Sydney NSW   
13                Sydney NSW   
14                Sydney NSW   

                                                title  \
0                       Junior Data Analyst/Scientist   
1                               Senior Data Scientist   
2                                  Research Scientist   
3                      Data Science Research Engineer   
4                             Data Scientist (Sydney)   
5                                      Data Scientist   
6                                      Data Scientist   
7                                      Data Scientist   
8                                    Data & Analytics   
9                                      Data Scientist   
10          Data Scientist, Google Technical Services   
11  Junior Quantitative Researcher - Machine Learning   
12  Google Cloud Platform Engineer - Big Data and ...   
13                 Data & Analytics - Leadership role   
14                          Data Scientist (Python/R)   

                                             company  \
0        International Institute of Data & Analytics   
1                                         Amazon.com   
2                                         Amazon.com   
3                               Macquarie University   
4                                             C3 IoT   
5                                         BuildingIQ   
6                                     Fortune Select   
7                             Black Swallow Boutique   
8                                               KPMG   
9                                         SAI Global   
10                                            Google   
11                                     Akuna Capital   
12                                          Deloitte   
13                                              KPMG   
14                               Correlate Resources   

                                         salary  \
0                                            NA   
1                                            NA   
2                                            NA   
3                    $100,706 - $112,058 a year   
4                                            NA   
5                                            NA   
6                                            NA   
7                                            NA   
8                                            NA   
9                                            NA   
10                                           NA   
11                                           NA   
12                                           NA   
13                                           NA   
14                   $100,000 - $120,000 a year   

                                              summary  
0                In data science and big data anal...  
1                A Senior Data Scientist will:. Ou...  
2    Data drives the development of new process; D...  
3                We are seeking a Data Science Res...  
4                In this capacity, you will partic...  
5                We are looking for a Data Scienti...  
6                Experience using and maintaining ...  
7                We are looking for an experienced...  
8                Using Data Science techniques and...  
9                Proficiency with Data Visualisati...  
10               Engage various stakeholders, asse...  
11               Akuna Capital is a young and boom...  
12               Experience working with recommend...  
13               Using Data Science techniques and...  
14               Work within a team of i

In [17]:
jobs0.shape

(17710, 5)

In [18]:
print(jobs0.duplicated().sum())

13305


In [19]:
print(jobs1.duplicated().sum())

7609


---
Using df.shape shows that there are 17,710 job records in the dataframe. However, when I check the number of duplicates, there are more than 13,305 duplicates for the first dataframe and 7,609 for the description dataframe. I will need to drop these duplicates.

----

In [20]:
jobs1.head(50)

description
0    The International Institute of Data & Analyti...
1    Excited by using massive amounts of data to d...
2    Business today operates at the pinnacle of th...
3                                                  NA
4    In this capacity, you will participate in the...
5    Job Description | Location – Sydney, Australi...
6    Location: Sydney  Job Type: Permanent  Skills...
7    We are looking for an experienced Data Scient...
8    Be a key part of our growth & innovation stra...
9    At SAI Global, we make Intelligent Risk possi...
10   Google Technical Services: Professional Servi...
11   About Akuna: Akuna Capital is a young and boo...
12   About Deloitte:   Flexibility, opportunity an...
13   Leadership opportunity in consulting (Associa...
14   Our client is an industry pioneering Customer...
15   Do you want to create positive change in Aust...
16   The Company:  Alloc8 have been fortunate enou...
17   I am looking for a Data Analyst with at least...
18   SiteMinder is the chosen solution for over 28...
19   Our Business More than 700 professionals in o...
20   Calling all Senior Data Scientists!!!!  The t...
21   Our client is an industry pioneering Customer...
22   For over 15 years Quantium have combined the ...
23   About Us: BIS Oxford Economics, part of Oxfor...
24   Since 2011, General Assembly has transformed ...
25   We are looking for a Data Scientist that will...
26   Scientist - 00006173  Job Title: Scientist Jo...
27   The International Institute of Data & Analyti...
28   The Senior Reporting & Statistical Analyst ex...
29   Research Analyst - Australia # 107055 Austral...
30   Event Research Analyst - 0000650N  Ongoing, F...
31   Make Amazing Happen!  An outstanding opportun...
32   Overview Our Quantitative Research team works...
33   Innovative, collaborative and positive team-o...
34   We are looking for intellectually curious tea...
35   About Us: EYC3 is the data and advanced analy...
36   Our client is an industry pioneering Customer...
37   Syphen is on the mission to unlock the potent...
38   Expedia  Are you excited about solving real w...
39   We are looking for intellectually curious tea...
40   Research Analyst - Australia # 107055 Austral...
41   Work type: Permanent Full time Location: Aust...
42   About Akuna: Akuna Capital is a young and boo...
43   About Akuna: Akuna Capital is a young and boo...
44   Requirements:  A minimum of Bachelor's degree...
45   PhD Applied Mathematics Proprietary trading -...
46   J.P. Morgan’s Corporate & Investment Bank is ...
47   Innovative, collaborative and positive team-o...
48   Here you'll get to work on cutting edge produ...
49   Front-End Developer - React, Angular, HTML, C...

In [21]:
jobs1.shape

(17710, 1)

---
I will now concatonate the two dataframes into one dataframe.

---

In [22]:
indeed_jobs = pd.concat([jobs0, jobs1], axis=1)

In [23]:
indeed_jobs.head(15)

location  \
0                 Sydney NSW   
1                 Sydney NSW   
2                 Sydney NSW   
3   Macquarie University NSW   
4                 Sydney NSW   
5                 Sydney NSW   
6                 Sydney NSW   
7             Alexandria NSW   
8                 Sydney NSW   
9                 Sydney NSW   
10                Sydney NSW   
11                Sydney NSW   
12                Sydney NSW   
13                Sydney NSW   
14                Sydney NSW   

                                                title  \
0                       Junior Data Analyst/Scientist   
1                               Senior Data Scientist   
2                                  Research Scientist   
3                      Data Science Research Engineer   
4                             Data Scientist (Sydney)   
5                                      Data Scientist   
6                                      Data Scientist   
7                                      Data Scientist   
8                                    Data & Analytics   
9                                      Data Scientist   
10          Data Scientist, Google Technical Services   
11  Junior Quantitative Researcher - Machine Learning   
12  Google Cloud Platform Engineer - Big Data and ...   
13                 Data & Analytics - Leadership role   
14                          Data Scientist (Python/R)   

                                             company  \
0        International Institute of Data & Analytics   
1                                         Amazon.com   
2                                         Amazon.com   
3                               Macquarie University   
4                                             C3 IoT   
5                                         BuildingIQ   
6                                     Fortune Select   
7                             Black Swallow Boutique   
8                                               KPMG   
9                                         SAI Global   
10                                            Google   
11                                     Akuna Capital   
12                                          Deloitte   
13                                              KPMG   
14                               Correlate Resources   

                                         salary  \
0                                            NA   
1                                            NA   
2                                            NA   
3                    $100,706 - $112,058 a year   
4                                            NA   
5                                            NA   
6                                            NA   
7                                            NA   
8                                            NA   
9                                            NA   
10                                           NA   
11                                           NA   
12                                           NA   
13                                           NA   
14                   $100,000 - $120,000 a year   

                                              summary  \
0                In data science and big data anal...   
1                A Senior Data Scientist will:. Ou...   
2    Data drives the development of new process; D...   
3                We are seeking a Data Science Res...   
4                In this capacity, you will partic...   
5                We are looking for a Data Scienti...   
6                Experience using and maintaining ...   
7                We are looking for an experienced...   
8                Using Data Science techniques and...   
9                Proficiency with Data Visualisati...   
10               Engage various stakeholders, asse...   
11               Akuna Capital is a young and boom...   
12               Experience working with recommend...   
13               Using Data Science techniques and...   
14               Work wit

In [24]:
indeed_jobs.tail(15)

location  \
17695              Adelaide SA   
17696              Adelaide SA   
17697              Adelaide SA   
17698              Adelaide SA   
17699              Adelaide SA   
17700              Adelaide SA   
17701              Adelaide SA   
17702  Adelaide City Centre SA   
17703              Adelaide SA   
17704              Adelaide SA   
17705              Adelaide SA   
17706              Adelaide SA   
17707              Adelaide SA   
17708              Adelaide SA   
17709              Adelaide SA   

                                                 title  \
17695                      Customer Service Specialist   
17696                      Sales Agronomist (Regional)   
17697                             Animation Supervisor   
17698                                   VFX Supervisor   
17699                                    CG Supervisor   
17700                            Management Accountant   
17701                                     Data Analyst   
17702                      Data Scientist / Data Miner   
17703                                    CG Supervisor   
17704                              Senior Test Analyst   
17705                                   VFX Supervisor   
17706                      Sales Agronomist (Regional)   
17707  Senior Microsoft Data Insights Consultant - Ade   
17708                      Customer Service Specialist   
17709                            Management Accountant   

                              company salary  \
17695                  Talent Options     NA   
17696                    Farmers Edge     NA   
17697       Rising Sun Pictures (RSP)     NA   
17698       Rising Sun Pictures (RSP)     NA   
17699       Rising Sun Pictures (RSP)     NA   
17700                    EGM Partners     NA   
17701            Piper Talent Pty Ltd     NA   
17702                      Peoplebank     NA   
17703       Rising Sun Pictures (RSP)     NA   
17704                     TicToc Home     NA   
17705       Rising Sun Pictures (RSP)     NA   
17706                    Farmers Edge     NA   
17707                     Empired Ltd     NA   
17708                  Talent Options     NA   
17709                    EGM Partners     NA   

                                                 summary  \
17695               Accurate data entry skills. This ...   
17696               From seed selection to yield data...   
17697               Please review our Privacy Policy ...   
17698               Please review our Privacy Policy ...   
17699               Please review our Privacy Policy ...   
17700               Work collaboratively with the Fin...   
17701               Microsoft Data Insights developme...   
17702               Experience in working in Big Data...   
17703               Please review our Privacy Policy ...   
17704               You'll have a passion for breakin...   
17705               Please review our Privacy Policy ...   
17706               From seed selection to yield data...   
17707               Microsoft Data Insights developme...   
17708               Accurate data entry skills. This ...   
17709               Work collaboratively with the Fin...   

                                             description  
17695   Customer Services Specialist  Our client is s...  
17696   Farmers Edge is a global leader in decision a...  
17697   Reporting directly to CG and VFX Supervisors,...  
17698   The VFX Supervisor is required to supervise, ...  
17699   The CG Supervisor will supervise, monitor, di...  
17700   We are resruiting a Senior/Management Account...  
17701   About my client A big company with a small fe...  
17702   Peoplebank is a preferred supplier to number ...  
17703   The CG Supervisor will supervise, monitor, di...  
17704   about TicToc The World's First Instant Home L...  
17705   The VFX Supervisor is required to supervise, ...  
17706   Farmers Edge is a global leader in decision a...  
17707   About Empired: The largest Asia Pacific ded

In [25]:
indeed_jobs.shape

(17710, 6)

---
I will now drop the duplicates. As the jobs0 dataframe has a lot more duplicates than jobs1, I will drop the duplicates that are found in the jobs0 dataframe to make sure all duplicates are removed.

---

In [26]:
indeed_jobs.drop_duplicates(['location', 'title', 'company', 'salary', 'summary'], inplace=True)

In [27]:
indeed_jobs.shape

(4405, 6)

---
Having removed the duplicates, there are 4,405 unique records remaining.

---

I have now finished the scrape and dropped the duplicates. I can save the data in csv format for future EDA.

---

In [29]:
indeed_jobs.to_csv('indeed_jobs.csv', index=False)

---
[End of Scraping (1). See file(2) for EDA]